cpp/include/raft/linalg/map.cuh

```diff
@@ -22,119 +22,74 @@
- * @tparam Args additional parameters
- * @tparam OutType data-type in which the result will be stored
- * @param out the output of the map operation (assumed to be a device pointer)
- * @param len number of elements in the input array
- * @param map the device-lambda
- * @param stream cuda-stream where to launch this kernel
- * @param in the input array
- * @param args additional input arrays
- */
-template <typename InType,
-          typename MapOp,
-          typename IdxType = std::uint32_t,
-          int TPB          = 256,
-          typename OutType = InType,
-          typename... Args>
-void map_k(
-  OutType* out, IdxType len, MapOp map, cudaStream_t stream, const InType* in, Args... args)
-{
-  detail::mapImpl<InType, OutType, IdxType, MapOp, TPB, Args...>(
-    out, len, map, stream, in, args...);
-}
+namespace raft::linalg {
 
 /**
  * @defgroup map Mapping ops
  * @{
  */
 
 /**
- * @brief CUDA version of map
- * @tparam InType data-type for math operation of type raft::device_mdspan
- * @tparam MapOp the device-lambda performing the actual operation
- * @tparam TPB threads-per-block in the final kernel launched
- * @tparam OutType data-type of result of type raft::device_mdspan
- * @tparam Args additional parameters
- * @param[in] handle raft::device_resources
- * @param[in] in the input of type raft::device_mdspan
- * @param[out] out the output of the map operation of type raft::device_mdspan
- * @param[in] map the device-lambda
- * @param[in] args additional input arrays
+ * @brief Map a function over zero or more input mdspans.
+ *
+ * Usage example:
+ * @code{.cpp}
+ *  auto input = raft::make_device_vector<int>(res, n);
+ *  ... fill input ..
+ *  auto squares = raft::make_device_vector<int>(res, n);
+ *  raft::linalg::map_offset(res, squares.view(), raft::sq_op{}, input.view());
+ * @endcode
+ *
+ * @tparam OutType data-type of the result (device_mdspan)
+ * @tparam Func the device-lambda performing the actual operation
+ * @tparam InTypes data-types of the inputs (device_mdspan)
+ *
+ * @param[in] res raft::device_resources
+ * @param[out] out the output of the map operation (device_mdspan)
+ * @param[in] f device lambda
+ * @param[in] ins the inputs (each of the same size as the output) (device_mdspan)
  */
-template <typename InType,
-          typename MapOp,
-          typename OutType,
-          int TPB = 256,
-          typename... Args,
-          typename = raft::enable_if_input_device_mdspan<InType>,
-          typename = raft::enable_if_output_device_mdspan<OutType>>
-void map(raft::device_resources const& handle, InType in, OutType out, MapOp map, Args... args)
+template <typename OutType,
+          typename Func,
+          typename... InTypes,
+          typename = raft::enable_if_output_device_mdspan<OutType>,
+          typename = raft::enable_if_input_device_mdspan<InTypes...>>
+void map(const raft::device_resources& res, OutType out, Func f, InTypes... ins)
```

Comment:
I really like this new function a lot and I'm really happy to see so many functions getting consolidated and using the same guts underneath. 

The concern that I have about this new API is that the ordering of the arguments breaks the consistency that we have been using across the new mdspan APIs.

We have been trying to follow this ordering:
1. Handle
2. Params struct (if it exists)
3. Inputs
4. Outputs
5. Any additional single params

This new signature puts the inputs after the outputs. To keep consistency, we could always swap the outputs and inputs across all the APIs... but that would be quite a change to make and would probably require another deprecation round if we did it right.

I do like the simplicity of the variadic templates, which is what makes this so hard. Another option could be to accept a std::vector of inputs like we do in the knn API, but then that would make it much harder to allow the types to differ across inputs.

Any other options you can think of? Again, I'm really happy with this change but would also like to try and keep the APIs as consistent as possible. I suppose if we squint our eyes enough, another perspective could be to look at the inputs individually as "additional parameters", but that's also kind of awkward.

Reply:
Thanks for the suggestions!

I don't think there is an easy way out of this situation. On the one hand, it makes sense to move the outputs to the beginning of the signatures, because it's very unlikely that we may need to use template parameter packs for them. On the other hand, I can imagine the parameter packs being used either for inputs or for extra single params (although less likely), hence changing the order to be always inputs-last does not guarantee to solve this forever.
Besides, there is at least one more such controversial yet useful function: [matrix/linewise_op](https://github.com/rapidsai/raft/blob/b12fc82be4c06cd157925ef14411fa2ee9cf0d07/cpp/include/raft/matrix/linewise_op.cuh#L65-L70), which has its mdspan inputs in the beginning as well as in the end and single parameters in the middle.

In my opinion, it's ok to have a little compromise to consistency for the few parameter-pack public functions we have but keep it strict where possible.

Perhaps, another option is to add a few more fixed overloads to map:
```
void map(const raft::device_resources& res, InType1 in1, OutType out, Func f);
void map(const raft::device_resources& res, InType1 in1, InType2 in2, OutType out, Func f);
void map(const raft::device_resources& res, InType1 in1, InType2 in2, InType2 in3, OutType out, Func f);
```
In theory, thanks to the lambda at the end the generic `map` version should not participate in instance resolution, and the type inference should work fine.

Reply:
Yeah, it's a hard problem for sure- I recall doing the mdspan for linewise_op and thinking something very similar (like squinting my eyes and trying to think of the additional vectors as params). 

I like the overloads too, and I think 3 of them could probably be enough to cover most cases.

As a middle ground, what would you think of doing something like this? 

```
map(handle, in, out, func, additional_in...)? 
```

I know it slightly complicates the internal logic, but in my perspective it stays true to our consistent arg ordering and it turns the `additional_in` into an additional params argument (without having to squint as much to justify it). Then in the case where the user wants zero inputs, we only need the one additional overload. I suppose we could do something similar here for linewise_op as well.

Reply:
> As a middle ground, what would you think of doing something like this?
NB: `map` can have zero input vectors (then it plays as `fill` or `range`)/

Reply:
(removed the previous remark, which I didn't notice on my phone was repeating what you've already said).

I don't know.. I don't like putting the different meanings into the first and the rest of the arguments, when they are semantically equal.

Can we just slightly modify the signature rule to state the layout as you declared it PLUS one exception that _if whatever group of arguments contains a parameter pack, the whole group goes last_? That is, if a part of the inputs is a parameter pack, all inputs pop to the end of the signature; if a part of the outputs is a parameter pack, then all outputs go after the single-parameter args, and so on.

Reply:
> Can we just slightly modify the signature rule to state the layout as you declared it PLUS one exception that if whatever group of arguments contains a parameter pack, the whole group goes last? That is, if a part of the inputs is a parameter pack, all inputs pop to the end of the signature; if a part of the outputs is a parameter pack, then all outputs go after the single-parameter args, and so on.

I'm okay with that, but I think I'm more okay with that if we do it in concert w/ this change as well:

```
void map(const raft::device_resources& res, InType1 in1, OutType out, Func f);
void map(const raft::device_resources& res, InType1 in1, InType2 in2, OutType out, Func f);
void map(const raft::device_resources& res, InType1 in1, InType2 in2, InType2 in3, OutType out, Func f);
```